# 重试

### 重试时间

当消费者消费异常时，框架会进行重试，默认会重试五次，按照 2 作为指数设置重试时间间隔。

第一次失败后，间隔两秒重试，第二次失败后，间隔四秒，下次是 8、16、32 秒。

默认重试策略代码：

```csharp
public class DefaultRetryPolicyFactory : IRetryPolicyFactory
{
    private readonly ILogger<DefaultRetryPolicyFactory> _logger;

    public DefaultRetryPolicyFactory(ILogger<DefaultRetryPolicyFactory> logger)
    {
        _logger = logger;
    }

    public virtual Task<AsyncRetryPolicy> CreatePolicy(string queue)
    {
        // Create a retry policy.
        // 创建重试策略.
        var retryPolicy = Policy
            .Handle<Exception>()
            .WaitAndRetryAsync(
                retryCount: 5,
                sleepDurationProvider: retryAttempt => TimeSpan.FromSeconds(Math.Pow(2, retryAttempt)),
                onRetry: async (Exception exception, TimeSpan timeSpan, int retryCount, Context context) =>
                {
                    _logger.LogDebug("Retry execution event,queue [{Queue}],retry count [{RetryCount}],timespan [{TimeSpan}]", queue, retryCount, timeSpan);
                    await FaildAsync(queue, exception, timeSpan, retryCount, context);
                });

        return Task.FromResult(retryPolicy);
    }

    public virtual Task FaildAsync(string queue, Exception ex, TimeSpan timeSpan, int retryCount, Context context)
    {
        return Task.CompletedTask;
    }
}
```



你可以通过实现 IRetryPolicyFactory 接口，替换默认的重试策略服务服务。

```
services.AddSingleton<IRetryPolicyFactory, DefaultRetryPolicyFactory>();
```



可以根据队列创建不同的重试策略。



### 重试机制

```csharp
    [Consumer("web1", Qos = 1 , RetryFaildRequeue = true)]
    public class MyConsumer : IConsumer<TestEvent>
    {
        private  int _retryCount = 0;
        // 消费
        public async Task ExecuteAsync(EventBody<TestEvent> message)
        {
            Console.WriteLine($"执行 {message.Body.Id} 第几次：{_retryCount} {DateTime.Now}");
            _retryCount++;
            throw new Exception("1");
        }

        // 每次失败时被执行
        public async Task FaildAsync(Exception ex, int retryCount, EventBody<TestEvent>? message)
        {
            Console.WriteLine($"重试 {message.Body.Id} 第几次：{retryCount} {DateTime.Now}");
            await Task.CompletedTask;
        }


        // 最后一次失败时执行
        public async Task<bool> FallbackAsync(EventBody<TestEvent>? message)
        {
            Console.WriteLine($"执行 {message.Body.Id} 补偿 {DateTime.Now}");
            return true;
        }
    }
}
```



首先会执行 `IConsumer<TEvent>.ExecuteAsync()` 或 `IEventMiddleware<TEvent>.ExecuteAsync()` 消费消息，如果代码执行失败，则会立即触发 `FaildAsync()` 函数。

一段时间间隔后，接着会重新执行消费。

比如默认重试机制是重试五次，那么最终 `IConsumer<TEvent>.ExecuteAsync()` 或 `IEventMiddleware<TEvent>.ExecuteAsync()` 都会被执行 6次，一次正常消费和五次重试消费。



FaildAsync 会在每次重试后执行，`Exceotion ex` 表示执行 `ExecuteAsync` 时出现的异常。

retryCount 表示当前执行 ExecuteAsync 的次数。



FallbackAsync 方法会在最后一次重试失败后被调用，该函数要放回一个 bool 类型。



当多次重试失败后，框架会调用 FallbackAsync 方法，如果该方法放回 true，那么框架会认为虽然 `ExecuteAsync()` 执行失败，但是通过 `FallbackAsync` 已经补偿好了，该消息会被当做正常完成消费，框架会向 RabbitMQ 服务器发送 ACK，接着消费下一条消息。



如果 `FallbackAsync ` 返回 true，框架会认为该消息彻底失败，如果设置了 `RetryFaildRequeue = true`，那么该条消息会被放回消息队列，等待下一次消费。否则该条消息会被直接丢弃。
